### ISSUE TITLE
MKL-DNN uses only single-core on Knight Landing processor

### ISSUE THREAD

**Issue by GD06** [01org/mkl-dnn#96]
I am testing the performance of MKL-DNN on the many-core processor Knight Landing 7250 for the convolution forward operation. However, the MKL-DNN runs too slow. I monitor the CPU utilization by both the ```htop``` and ```top``` and I am surprised to find that MKL-DNN uses only one core, which doesn't take advantage of the many-core architecture. As we know, Knight Landing is equipped with many processing cores. I don't set the environment variable ```OMP_NUM_THREADS``` and don't have any idea about what environment variables I should set or unset.

The compilation options can be found in the [Makefile](https://gist.github.com/414ae9c066de390076c2f564c8d5cafc.git). The source codes are uploaded in the [mkldnn_conv.cpp](https://gist.github.com/14b4d91f6904e836689c2dad592e2a5a.git) also.

Thanks for the help!

**Comment by rsdubtso** [comment]
Which compiler do you use? KNL is one of the target platforms for MKL-DNN, and the code can definitely scale to a full chip. Does the behavior change if you do set `OMP_NUM_THREADS`?

**Comment by GD06** [comment]
@rsdubtso Thanks for your reply! I tried to set ```OMP_NUM_THREADS=272```, but it doesn't help. I compile it with g++-5.4.

**Comment by rsdubtso** [comment]
Can you please post the output of the test on your system running with environment `OMP_DISPLAY_ENV=TRUE`?

**Comment by GD06** [comment]
Hi @rsdubtso ,
Thank you for your suggestions. For reproducing my issue, I upload the [main.cpp](https://gist.github.com/f2d54e1eb173d434fc04fe65de8b6455.git) and the [im2col_mkl.cpp](https://gist.github.com/b1f00b385c85544547d1a083c0eeda6a.git) so that you can compile with the Makefile together with the mkldnn_conv.cpp I uploaded in the issue description. This program is used to test the performance of MKL-DNN. After compilation, the size of input tensors can be configured by the command line arguments. Currently, I use the following command to execute the program:

```
./main 1 32 224 224 3 3 32
``` 

After enabling ```OMP_DISPLAY_ENV=TRUE```, the outputs are as following:

```
OPENMP DISPLAY ENVIRONMENT BEGIN
  _OPENMP = '201307'
  OMP_DYNAMIC = 'FALSE'
  OMP_NESTED = 'FALSE'
  OMP_NUM_THREADS = '272'
  OMP_SCHEDULE = 'DYNAMIC'
  OMP_PROC_BIND = 'FALSE'
  OMP_PLACES = ''
  OMP_STACKSIZE = '140683439707464'
  OMP_WAIT_POLICY = 'PASSIVE'
  OMP_THREAD_LIMIT = '4294967295'
  OMP_MAX_ACTIVE_LEVELS = '2147483647'
  OMP_CANCELLATION = 'FALSE'
  OMP_DEFAULT_DEVICE = '0'
OPENMP DISPLAY ENVIRONMENT END

OPENMP DISPLAY ENVIRONMENT BEGIN
   _OPENMP='201511'
  [host] OMP_CANCELLATION='FALSE'
  [host] OMP_DEFAULT_DEVICE='0'
  [host] OMP_DISPLAY_ENV='TRUE'
  [host] OMP_DYNAMIC='FALSE'
  [host] OMP_MAX_ACTIVE_LEVELS='2147483647'
  [host] OMP_MAX_TASK_PRIORITY='0'
  [host] OMP_NESTED='FALSE'
  [host] OMP_NUM_THREADS: value is not defined
  [host] OMP_PLACES: value is not defined
  [host] OMP_PROC_BIND='intel'
  [host] OMP_SCHEDULE='static'
  [host] OMP_STACKSIZE='4M'
  [host] OMP_THREAD_LIMIT='2147483647'
  [host] OMP_WAIT_POLICY='PASSIVE'
OPENMP DISPLAY ENVIRONMENT END
```

It seems that the execution of MKL matrix multiplication can use many-core normally, but MKL-DNN is executed with another OPENMP environments. Do you have any idea about this issue? I am a beginner with OpenMP. I suspect the problem comes from the Makefile configuration, but I don't have any idea to fix it.

If you have seconds to reproduce my issue while encountering any problem, I will provide all my source codes with system environments as far as I can. Thank you again for your time!

**Comment by rsdubtso** [comment]
Thanks, I'll look into this.

**Comment by rsdubtso** [comment]
I've just realized that you are mixing intel (`libiomp5.so`) and gnu (`libgomp.so`) threading libraries. You need to remove `-liomp5` from the link line and replace `libmkl_intel_thread.a` with `libmkl_gnu_thread.a`. Alternatively, if your own code does not use OpenMP at all, you can just remove the `-fopenmp` from the compiler flags.

Sorry this took so long. I should've looked at the Makefile at the very start.

**Comment by GD06** [comment]
Hi @rsdubtso ,

Thank you for your time! I edit the Makefile as you suggested. However, it doesn't work. I attach the updated [Makefile](https://gist.github.com/414ae9c066de390076c2f564c8d5cafc.git) into the Gist already. The performance results of rerunning the program after the compilation don't change. Besides, when I monitor the htop, I found that only single-core is used still. The following are the outputs after executing the command ```OMP_DISPLAY_ENV=TRUE ./main 1 32 224 224 3 3 32```. The biggest different thing I can notice is that this time, the OpenMP environments are displayed only once. I am sorry that I still don't have any idea.

```
OPENMP DISPLAY ENVIRONMENT BEGIN
   _OPENMP='201511'
  [host] OMP_CANCELLATION='FALSE'
  [host] OMP_DEFAULT_DEVICE='0'
  [host] OMP_DISPLAY_ENV='TRUE'
  [host] OMP_DYNAMIC='FALSE'
  [host] OMP_MAX_ACTIVE_LEVELS='2147483647'
  [host] OMP_MAX_TASK_PRIORITY='0'
  [host] OMP_NESTED='FALSE'
  [host] OMP_NUM_THREADS: value is not defined
  [host] OMP_PLACES: value is not defined
  [host] OMP_PROC_BIND='intel'
  [host] OMP_SCHEDULE='static'
  [host] OMP_STACKSIZE='4M'
  [host] OMP_THREAD_LIMIT='2147483647'
  [host] OMP_WAIT_POLICY='PASSIVE'
OPENMP DISPLAY ENVIRONMENT END
```

**Comment by GD06** [comment]
Hi @rsdubtso 

I just realized where is the problem. Thank for your pointing me out that I mix the thread libraries! I use larger problem size and run the convolution operation repeatedly then I can observe the MKL-DNN convolution using 68 threads in the htop. Previously, I used too small problem size so the major time was spent on the pre-processing stage, either preparing data or creating the descriptor. That's why I observed only single-core used for the processing. 

Thank you again for helping me a lot!

Thanks,
Xinfeng
